addons/hr_expense/models/hr_expense.py

```diff
@@ -197,6 +201,59 @@ def _compute_employee_id(self):
             for expense in self:
                 expense.employee_id = self.env.user.with_company(expense.company_id).employee_id
 
+    @api.depends('employee_id', 'product_id', 'total_amount')
+    def _compute_duplicate_expense_ids(self):
+        self.duplicate_expense_ids = False
+
+        expenses = self.filtered(lambda e: e.employee_id and e.product_id and e.total_amount)
+        if expenses.ids:
+            duplicates_query = """
+                WITH expenses AS (
+                    SELECT employee_id, product_id, date, total_amount, company_id, currency_id
+                      FROM hr_expense
+                     WHERE id IN %(expense_ids)s
+                     GROUP BY employee_id, product_id, date, total_amount, company_id, currency_id)
+              SELECT ARRAY_AGG(he.id)
+                FROM hr_expense AS he
+               INNER JOIN expenses AS ex
+                  ON he.employee_id = ex.employee_id
+                 AND he.product_id = ex.product_id
+                 AND he.date = ex.date
+                 AND he.total_amount = ex.total_amount
+                 AND he.company_id = ex.company_id
+                 AND he.currency_id = ex.currency_id
+               GROUP BY he.employee_id, he.product_id, he.date, he.total_amount, he.company_id, he.currency_id
+              HAVING COUNT(id) > 1
+            """
+            self.env.cr.execute(duplicates_query, {
+                'expense_ids': tuple(expenses.ids),
+            })
+            duplicates = [x[0] for x in self.env.cr.fetchall()]
+
+            for ids in duplicates:
+                exp = expenses.filtered(lambda e: e.id in ids)
+                exp.duplicate_expense_ids = [(6, 0, ids)]
+                expenses = expenses - exp
+
+        new_expenses = expenses.filtered(lambda e: not e.id)
+        if new_expenses:
+            domain = []
+            for expense in new_expenses:
+                domain = expression.OR([domain, [
+                    ('employee_id', '=', expense.employee_id.id),
+                    ('product_id', '=', expense.product_id.id),
+                    ('date', '=', expense.date),
+                    ('total_amount', '=', expense.total_amount),
+                    ('company_id', '=', expense.company_id.id),
+                    ('currency_id', '=', expense.currency_id.id),
+                ]])
+            duplicates = self.env['hr.expense'].search(domain)
+            for expense in new_expenses:
+                expense.duplicate_expense_ids = duplicates.filtered(lambda e: e.employee_id == expense.employee_id
+                                                                    and e.product_id == expense.product_id and e.date == expense.date
+                                                                    and e.total_amount == expense.total_amount
+                                                                    and e.company_id == expense.company_id and e.currency_id == expense.currency_id)
+
```

Comment:
Just posting for the posterity of our informal discussion.
Your are doing an SQL request for existing expenses. For new expenses, you use a domain. I don't think the number of expense record is big enough to slow down the experience but if that was the case, we could need think of an SQL query for both type of records. Before that, we can keep your current code. We will see what @tivisse have in mind :-)